This patch contains backports of the following securiy and stability fixes:

https://github.com/miniupnp/miniupnp/commit/7aeb624b44f86d335841242ff427433190e7168a
https://github.com/miniupnp/miniupnp/commit/a0573e251817ec090a8c9f9f41b56d720c835a6c
https://github.com/miniupnp/miniupnp/commit/e1b4f25bba4bd5146bfb950b3c5c2ce711a43e5f

diff -r f9ab835ddd56 miniupnpc/minixml.c
--- a/miniupnpc/minixml.c	Fri Jun 12 16:28:32 2020 +0200
+++ b/miniupnpc/minixml.c	Fri Jun 12 16:54:11 2020 +0200
@@ -161,7 +161,8 @@
 						if (p->xml >= p->xmlend)
 							return;
 					}
-					if(memcmp(p->xml, "<![CDATA[", 9) == 0)
+					/* CDATA are at least 9 + 3 characters long : <![CDATA[ ]]> */
+					if((p->xmlend >= (p->xml + (9 + 3))) && (memcmp(p->xml, "<![CDATA[", 9) == 0))
 					{
 						/* CDATA handling */
 						p->xml += 9;
diff -r f9ab835ddd56 miniupnpc/upnpreplyparse.c
--- a/miniupnpc/upnpreplyparse.c	Fri Jun 12 16:28:32 2020 +0200
+++ b/miniupnpc/upnpreplyparse.c	Fri Jun 12 16:54:11 2020 +0200
@@ -77,6 +77,7 @@
 	if(strcmp(data->curelt, "NewPortListing") == 0)
 	{
 		/* specific case for NewPortListing which is a XML Document */
+		free(data->portListing);
 		data->portListing = malloc(l + 1);
 		if(!data->portListing)
 		{
@@ -104,9 +105,7 @@
                struct NameValueParserData * data)
 {
 	struct xmlparser parser;
-	data->l_head = NULL;
-	data->portListing = NULL;
-	data->portListingLength = 0;
+	memset(data, 0, sizeof(struct NameValueParserData));
 	/* init xmlparser object */
 	parser.xmlstart = buffer;
 	parser.xmlsize = bufsize;
